bpo-30699: Improve example on tzinfo instances

mariocj89 opened the pull request:
Move from GMTX to TZX when naming the classes, as GMT1 might be rather confusing as seen in the reported issue.
In addition, move to UTC over GMT and improve the tzname implementation.

<!-- issue-number: bpo-30699 -->
https://bugs.python.org/issue30699
<!-- /issue-number -->

mariocj89 committed:
Improve example on tzinfo instances

Move from GMTX to TZX when naming the classes, as GMT1 might be rather
confusing as seen in the reported issue.
In addition, move to UTC over GMT and improve the tzname implementation.

vstinner reviewed (approved):
LGTM.

abalkin reviewed (dismissed):
Let's make the examples [PEP 495](https://www.python.org/dev/peps/pep-0495) compliant.

mariocj89 committed:
Simplify datetime with tzinfo example

Move the example in the documentation to just use timezone.utc and a
user defined Kabul timezone rather than having two user defined
timezones with DST.

Kabul timezone is still interesting as it changes its offset but not
based on DST. This is more accurate as the previous example was missing
information about the fold attribute. Additionally, implementing the fold
attribute was rather complex and probably not relevant enough for the
section "datetime with tzinfo".

abalkin commented on Doc/library/datetime.rst:
```diff
@@ -1316,56 +1316,58 @@ Examples of working with datetime objects:
+    ...
     ...     def tzname(self,dt):
-    ...         return "GMT +2"
+    ...         return "Asia/Kabul"
```
I think this should return UTC+04:30 / UTC+04 depending on the time relative to the move date.

mariocj89 replied:
From the docs:

> Return the time zone name corresponding to the datetime object dt, as a string. Nothing about string names is defined by the datetime module, and there’s no requirement that it mean anything in particular. For example, “GMT”, “UTC”, “-500”, “-5:00”, “EDT”, “US/Eastern”, “America/New York” are all valid replies

If the method is to return the timezone *name* I think it is probably better to have Kabul rather than the string representation of the offset. That said, both dateutil and pytz are returning the string representation of the offset, not sure why.

Happy to change it if you still think it is more correct.

Maybe @pganssle can reason why it is the offset rather than the name in dateutil.

abalkin replied:
Traditionally, tzname() is returning something similar to strftime('%Z').  In the case of Kabul, this means

```
$ TZ=Asia/Kabul date -r 1000000000 +%Z
+0430
```
for recent times, and
```
$ TZ=Asia/Kabul date -r -1000000000 +%Z
+04
```
for times before the transition.  Paul Eggert, has been leading a [crusade on "invented" timezone abbreviations](https://mm.icann.org/pipermail/tz/2017-February/024837.html) recently resulting in %Z format becoming useless for many parts of the World.  I still believe tzname() should return a time-dependent abbreviation rather than a permanent zone name.

mariocj89 replied:
OK! Updated via https://github.com/python/cpython/pull/4290/commits/56e7adfe703d8eb0b165eeeb131d27d86c9ccbee

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -1316,56 +1316,61 @@ Examples of working with datetime objects:
+    datetime.timedelta(0, 14400)
+    >>> # Datetime after the change
+    >>> dt2 = datetime(2006, 6, 14, 13, 0, tzinfo=tz1)
     >>> dt2.utcoffset()
```
Can we change this to:

```python
>>> print(dt.utcoffset())
4:30:00
```

It's a very minor thing, but to me the standard `repr` for `timedelta` is completely unreadable.

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -1316,56 +1316,61 @@ Examples of working with datetime objects:
-    ...             return timedelta(0)
-    ...     def tzname(self,dt):
-    ...         return "GMT +2"
+    ...             return "+04"
```
I would mildly prefer `"+04:00"` here for consistency, but I don't care strongly about this, it's just a documentation example and I do realize that @abalkin suggested this label.

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -1316,56 +1316,61 @@ Examples of working with datetime objects:
-    >>> from datetime import timedelta, datetime, tzinfo
-    >>> class GMT1(tzinfo):
+    >>> from datetime import timedelta, datetime, tzinfo, timezone
+    >>> class KabulTz(tzinfo):
```
If you add:

```python
def __repr__(self):
    return f"{self.__class__.__name__}()"
```
The later examples where you attach this to a datetime will be easier to read, e.g.:

```python
   >>> dt2     # doctest: +ELLIPSIS
   datetime.datetime(2006, 6, 14, 13, 0, tzinfo=KabulTz())
```

You may also consider adding an `__str__` that returns the tzdata IANA key (though I believe this is used nowhere in your examples later):

```python
def __str__(self):
    return "Asia/Kabul"
```

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -1591,26 +1596,25 @@ Instance methods:
     ...         return timedelta(0)
     ...     def tzname(self,dt):
-    ...         return "Europe/Prague"
+    ...         return "UTC+1"
```
Super annoyingly, [POSIX-style offset names](https://github.com/dateutil/dateutil/issues/70#issuecomment-88982122) have an inverted sense, so `dateutil.parser.parse` (and I think `TZ` variables) would interpret `UTC+1` as the same as `-01:00`. Of course, many people use this in the opposite sense, which gives rise to some confusion.

I recommend bypassing the whole thing by returning "+01:00" here.

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -1591,26 +1596,25 @@ Instance methods:
 
     >>> from datetime import time, tzinfo, timedelta
-    >>> class GMT1(tzinfo):
+    >>> class TZ1(tzinfo):
```
As with above, a `__repr__` would be a good idea here.

pganssle reviewed (changes requested):
Almost all of my comments here are completely optional and I think this can be merged whether they are incorporated or not.

The only "hard" change that I recommend is the use of `"+01:00"` instead of `"UTC+1"` in the `TZ1` class.

pganssle reviewed (approved):
@vstinner No problem, approved. I think this change is clearly an improvement over the status quo.

I think documentation backports are only going back to 3.7 now anyway, but in case they aren't, this relies on PEP 495, so it can't be backported any further than 3.6. I think it's fine to backport to 3.7 unless @mariocj89 disagrees.
